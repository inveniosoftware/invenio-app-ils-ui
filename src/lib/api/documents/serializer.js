import { recordResponseSerializer } from '@api/utils';

const DocumentSerializers = {
  responseSerializer: function(hit) {
    return recordResponseSerializer(hit);
  },
};

export const documentSerializer = {
  fromJSON: DocumentSerializers.responseSerializer,
};
