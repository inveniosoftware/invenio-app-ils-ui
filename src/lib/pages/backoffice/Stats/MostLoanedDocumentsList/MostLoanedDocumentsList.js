import SearchResultsList from '@modules/SearchControls/SearchResultsList';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Qs from 'qs';
import { Grid, List, Header } from 'semantic-ui-react';
import { DatePicker } from '@components/DatePicker';
import { Loader } from '@components/Loader';
import { Error } from '@components/Error';
import { BackOfficeRoutes } from '@routes/urls';
import { circulationStatsApi } from '@api/stats/circulationStats';
import { loanApi } from '@api/loans';
import { ExportSearchResults } from '@components/backoffice/ExportSearchResults';
import { invenioConfig } from '@config';
import {
  DocumentListEntry,
  DocumentStats,
} from '@modules/Document/backoffice/DocumentList';

export default class MostLoanedDocumentsList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      fromDate: '',
      toDate: '',
    };
  }

  componentDidMount() {
    this.fetchDocuments();
  }

  get subtitle() {
    const { fromDate, toDate } = this.state;
    const { data } = this.props;
    const total = data.total;
    const msg = `Showing the ${total} most checked-out documents`;

    if (fromDate && toDate) {
      return `${msg} between ${fromDate} and ${toDate}.`;
    } else if (fromDate) {
      return `${msg} starting from ${fromDate}.`;
    } else if (toDate) {
      return `${msg} ending at ${toDate}.`;
    } else {
      return `${msg}.`;
    }
  }

  fetchDocuments() {
    const { fromDate, toDate } = this.state;
    const { fetchMostLoanedDocuments } = this.props;
    fetchMostLoanedDocuments(fromDate, toDate);
  }

  handleFromDateChange = value => {
    this.setState({ fromDate: value }, this.fetchDocuments);
  };

  handleToDateChange = value => {
    this.setState({ toDate: value }, this.fetchDocuments);
  };

  renderDateRangePicker() {
    const { fromDate, toDate } = this.state;

    return (
      <>
        <Header as="h3">
          Pick range of dates
          <Header.Subheader>
            Choose between which dates to search the most loaned documents
          </Header.Subheader>
        </Header>

        <List horizontal>
          <List.Item>
            <List.Content>
              <DatePicker
                maxDate={toDate}
                placeholder="From Date"
                handleDateChange={this.handleFromDateChange}
              />
            </List.Content>
          </List.Item>

          <List.Item>
            <List.Content>
              <DatePicker
                minDate={fromDate}
                placeholder="To Date"
                handleDateChange={this.handleToDateChange}
              />
            </List.Content>
          </List.Item>
        </List>
      </>
    );
  }

  renderHeader = () => {
    const { fromDate, toDate } = this.state;

    return (
      <Grid columns={2}>
        <Grid.Column>
          <Header as="h3">
            Most Loaned Documents
            <Header.Subheader>{this.subtitle}</Header.Subheader>
          </Header>
        </Grid.Column>
        <Grid.Column textAlign="right">
          <ExportSearchResults
            onExportClick={(format, size) => {
              // build params
              const params = circulationStatsApi.getMostLoanedDocumentsParams(
                fromDate,
                toDate,
                size,
                format
              );
              const args = Qs.stringify(params);
              // build final url
              const exportUrl = `${circulationStatsApi.mostLoanedUrl}?${args}`;
              // open in a new tab
              window.open(exportUrl, '_blank');
            }}
          />
        </Grid.Column>
      </Grid>
    );
  };

  createLoansUrl = (hit, fromDate, toDate) => {
    const query = loanApi
      .query()
      .withDocPid(hit.metadata.pid)
      .withStartDate({ fromDate, toDate })
      .withState(
        invenioConfig.CIRCULATION.loanActiveStates.concat(
          invenioConfig.CIRCULATION.loanCompletedStates
        )
      )
      .qs();
    return BackOfficeRoutes.loansListWithQuery(query);
  };

  render() {
    const { data, isLoading, error } = this.props;
    const { fromDate, toDate } = this.state;

    const hitsWithLinks = data.hits.map(hit => {
      hit.metadata.loan_count_url = this.createLoansUrl(hit, fromDate, toDate);
      return hit;
    });
    return (
      <Grid columns={1}>
        <Grid.Column>{this.renderDateRangePicker()}</Grid.Column>
        <Grid.Column>{this.renderHeader()}</Grid.Column>
        <Grid.Column>
          <Loader isLoading={isLoading}>
            <Error error={error}>
              <SearchResultsList
                results={hitsWithLinks}
                ListEntryElement={DocumentListEntry}
                renderMiddleColumn={doc => (
                  <DocumentStats metadata={doc.metadata} />
                )}
              />
            </Error>
          </Loader>
        </Grid.Column>
      </Grid>
    );
  }
}

MostLoanedDocumentsList.propTypes = {
  fetchMostLoanedDocuments: PropTypes.func.isRequired,
  data: PropTypes.object.isRequired,
  isLoading: PropTypes.bool,
  error: PropTypes.object,
};

MostLoanedDocumentsList.defaultProps = {
  isLoading: false,
  error: null,
};
