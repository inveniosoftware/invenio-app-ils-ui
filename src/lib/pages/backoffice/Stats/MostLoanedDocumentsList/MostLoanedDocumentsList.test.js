import React from 'react';
import { shallow, mount } from 'enzyme';
import { Settings } from 'luxon';
import MostLoanedDocumentsList from './MostLoanedDocumentsList';
import { MemoryRouter } from 'react-router';
import * as testData from '@testData/documents.json';

jest.mock('@components/backoffice/ExportSearchResults');
jest.mock('@config');

Settings.defaultZoneName = 'utc';
const stringDate = '2018-01-01T11:05:00+01:00';

describe('MostLoanedDocumentsList tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  it('should load the component', () => {
    const component = shallow(
      <MostLoanedDocumentsList
        data={{ hits: [], total: 0 }}
        fetchMostLoanedDocuments={() => {}}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should fetch documents on mount', () => {
    const mockedFetchLoans = jest.fn();
    component = mount(
      <MostLoanedDocumentsList
        data={{ hits: [], total: 0 }}
        fetchMostLoanedDocuments={mockedFetchLoans}
      />
    );
    expect(mockedFetchLoans).toHaveBeenCalled();
  });

  it('should render documents', () => {
    const data = {
      hits: [
        {
          id: 1,
          updated: stringDate,
          created: stringDate,
          pid: 'doc1',
          metadata: {
            ...testData[0],
            pid: 'doc1',
            loan_count: 1,
            extension_count: 1,
            items: { total: 2 },
            eitems: { total: 2 },
            circulation: {
              past_loans_count: 1,
            },
            relations: {},
          },
        },
        {
          id: 2,
          updated: stringDate,
          created: stringDate,
          pid: 'doc2',
          metadata: {
            ...testData[1],
            pid: 'doc2',
            loan_count: 1,
            extension_count: 1,
            items: { total: 2 },
            eitems: { total: 2 },
            circulation: {
              past_loans_count: 1,
            },
            relations: {},
          },
        },
      ],
      total: 2,
    };
    component = mount(
      <MemoryRouter>
        <MostLoanedDocumentsList
          data={data}
          fetchMostLoanedDocuments={() => {}}
        />
      </MemoryRouter>
    );

    const rows = component.find('Item');
    expect(rows).toHaveLength(2);
  });
});
