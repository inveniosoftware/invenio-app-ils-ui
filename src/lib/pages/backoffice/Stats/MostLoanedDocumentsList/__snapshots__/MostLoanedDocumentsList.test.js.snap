// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MostLoanedDocumentsList tests should load the component 1`] = `
<Grid
  columns={1}
>
  <GridColumn>
    <Header
      as="h3"
    >
      Pick range of dates
      <HeaderSubheader>
        Choose between which dates to search the most loaned documents
      </HeaderSubheader>
    </Header>
    <List
      horizontal={true}
    >
      <ListItem>
        <ListContent>
          <Overridable(DatePicker)
            handleDateChange={[Function]}
            maxDate=""
            placeholder="From Date"
          />
        </ListContent>
      </ListItem>
      <ListItem>
        <ListContent>
          <Overridable(DatePicker)
            handleDateChange={[Function]}
            minDate=""
            placeholder="To Date"
          />
        </ListContent>
      </ListItem>
    </List>
  </GridColumn>
  <GridColumn>
    <Grid
      columns={2}
    >
      <GridColumn>
        <Header
          as="h3"
        >
          Most Loaned Documents
          <HeaderSubheader>
            Showing the 0 most checked-out documents.
          </HeaderSubheader>
        </Header>
      </GridColumn>
      <GridColumn
        textAlign="right"
      >
        <ExportSearchResults
          onExportClick={[Function]}
        />
      </GridColumn>
    </Grid>
  </GridColumn>
  <GridColumn>
    <Overridable(Loader)
      isLoading={false}
    >
      <Error
        FallbackComponent={[Function]}
        boundary={false}
        error={null}
        onUIError={null}
      >
        <SearchResultsList
          ListEntryElement={[Function]}
          renderMiddleColumn={[Function]}
          results={Array []}
        />
      </Error>
    </Overridable(Loader)>
  </GridColumn>
</Grid>
`;
